### PR TITLE
Extend JSON file reader to handle array of objects with resourceSpans

### LINKED ISSUE #2225: [Bug]: Importing OTLP "File Exporter" JSON fails
### What happened?

jaegertracing/all-in-one 1.55 fails to import JSON generated by OTEL File Exporter.

### Steps to reproduce

1. Spin up jaegertracing/all-in-one 1.55
2. Save [trace example JSON data](https://opentelemetry.io/docs/specs/otel/protocol/file-exporter/#examples) to `test.json`
```json
{"resourceSpans":[{"resource":{"attributes":[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]},"scopeSpans":[{"scope":{},"spans":[{"traceId":"","spanId":"","parentSpanId":"","name":"operationA","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","droppedAttributesCount":1,"events":[{"timeUnixNano":"1581452773000000123","name":"event-with-attr","attributes":[{"key":"span-event-attr","value":{"stringValue":"span-event-attr-val"}}],"droppedAttributesCount":2},{"timeUnixNano":"1581452773000000123","name":"event","droppedAttributesCount":2}],"droppedEventsCount":1,"status":{"message":"status-cancelled","code":2}},{"traceId":"","spanId":"","parentSpanId":"","name":"operationB","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","links":[{"traceId":"","spanId":"","attributes":[{"key":"span-link-attr","value":{"stringValue":"span-link-attr-val"}}],"droppedAttributesCount":4},{"traceId":"","spanId":"","droppedAttributesCount":1}],"droppedLinksCount":3,"status":{}}]}]}]}
{"resourceSpans":[{"resource":{"attributes":[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]},"scopeSpans":[{"scope":{},"spans":[{"traceId":"","spanId":"","parentSpanId":"","name":"operationA","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","droppedAttributesCount":1,"events":[{"timeUnixNano":"1581452773000000424","name":"event-with-attr","attributes":[{"key":"span-event-attr","value":{"stringValue":"span-event-attr-val"}}],"droppedAttributesCount":2},{"timeUnixNano":"1581452773000000424","name":"event","droppedAttributesCount":2}],"droppedEventsCount":1,"status":{"message":"status-cancelled","code":2}},{"traceId":"","spanId":"","parentSpanId":"","name":"operationB","startTimeUnixNano":"1581452772000000343","endTimeUnixNano":"1581452773000001089","links":[{"traceId":"","spanId":"","attributes":[{"key":"span-link-attr","value":{"stringValue":"span-link-attr-val"}}],"droppedAttributesCount":3},{"traceId":"","spanId":"","droppedAttributesCount":4}],"droppedLinksCount":2,"status":{}}]}]}]}
{"resourceSpans":[{"resource":{"attributes":[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]},"scopeSpans":[{"scope":{},"spans":[{"traceId":"","spanId":"","parentSpanId":"","name":"operationA","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","droppedAttributesCount":1,"events":[{"timeUnixNano":"1581452773000000826","name":"event-with-attr","attributes":[{"key":"span-event-attr","value":{"stringValue":"span-event-attr-val"}}],"droppedAttributesCount":2},{"timeUnixNano":"1581452773000000826","name":"event","droppedAttributesCount":2}],"droppedEventsCount":1,"status":{"message":"status-cancelled","code":2}},{"traceId":"","spanId":"","parentSpanId":"","name":"operationB","startTimeUnixNano":"1581452772000200521","endTimeUnixNano":"1581452773000004789","links":[{"traceId":"","spanId":"","attributes":[{"key":"span-link-attr","value":{"stringValue":"span-link-attr-val"}}],"droppedAttributesCount":5},{"traceId":"","spanId":"","droppedAttributesCount":2}],"droppedLinksCount":3,"status":{}}]}]}]}
{"resourceSpans":[{"resource":{"attributes":[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]},"scopeSpans":[{"scope":{},"spans":[{"traceId":"","spanId":"","parentSpanId":"","name":"operationA","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","droppedAttributesCount":1,"events":[{"timeUnixNano":"1581452773000010925","name":"event-with-attr","attributes":[{"key":"span-event-attr","value":{"stringValue":"span-event-attr-val"}}],"droppedAttributesCount":2},{"timeUnixNano":"1581452773000010925","name":"event","droppedAttributesCount":2}],"droppedEventsCount":1,"status":{"message":"status-cancelled","code":2}},{"traceId":"","spanId":"","parentSpanId":"","name":"operationB","startTimeUnixNano":"1581452772000011821","endTimeUnixNano":"1581452772000012924","links":[{"traceId":"","spanId":"","attributes":[{"key":"span-link-attr","value":{"stringValue":"span-link-attr-val"}}],"droppedAttributesCount":2},{"traceId":"","spanId":"","droppedAttributesCount":2}],"droppedLinksCount":5,"status":{}}]}]}]}
```
3. Drag `test.json` onto Jaeger's `Click or drag files to this area.`
4. Get error:
```
There was an error querying for traces:
Error parsing JSON: JSON.parse: unexpected non-whitespace character after JSON data at line 2 column 1 of the JSON data
```

### Expected behavior

I expect Jaeger to have loaded successfully loaded the Open Telemetry traces, per https://github.com/jaegertracing/jaeger/issues/4949

### Relevant log output

_No response_

### Screenshot

image

### Additional context

The existing code only expects there to be a single JSON object entry in the OTEL File Exporter JSON output, but the [file exporter specification](https://opentelemetry.io/docs/specs/otel/protocol/file-exporter/#json-lines-file) says that the is in JSON lines format (file contains multiple JSON serialized objects, with a new-line separating each).

This is evident when looking at the test data in the original PR, as it is only a single JSON object entry.
https://github.com/jaegertracing/jaeger/pull/5155/files

### Jaeger backend version

1.55

### SDK

_No response_

### Pipeline

Example of the opentelemetry collector configuration I use to capture OTEL traces to OTEL file-exporter format:

```yaml
extensions:
  zpages:
    endpoint: 0.0.0.0:55679

receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318

processors:
  batch:
  memory_limiter:
    # 75% of maximum memory up to 2G
    limit_mib: 1536
    # 25% of limit up to 2G
    spike_limit_mib: 512
    check_interval: 5s

exporters:
  file:
    path: /output/otel.json

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [memory_limiter, batch]
      exporters: [file]

  extensions: [zpages]
```

### Stogage backend

_No response_

### Operating system

_No response_

### Deployment model

_No response_

### Deployment configs

_No response_

### PR DESCRIPTION
**Pull request by Adarsh-jaiss**
## Which problem is this PR solving?
- Resolves #2225 

## Description of the changes
- This commit updates the JSON file reader function to handle an array of JSON objects containing 'resourceSpans' arrays. It now merges these arrays into a single 'resourceSpans' array before processing, ensuring compatibility with both single JSON objects and arrays of objects.

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: packages/jaeger-ui/src/utils/readJsonFile.tsx [hunk 1]
```diff
@@ -24,7 +24,21 @@ export default function readJsonFile(fileList: { file: File }): Promise<string>
...
         const traceObj = JSON.parse(reader.result);
-        if ('resourceSpans' in traceObj) {
+        if (Array.isArray(traceObj) && traceObj.every(obj => 'resourceSpans' in obj)) {
+          const mergedResourceSpans = traceObj.reduce((acc, obj) => {
+            acc.push(...obj.resourceSpans);
+            return acc;
+          }, []);
+
+          const mergedTraceObj = { resourceSpans: mergedResourceSpans };
+          JaegerAPI.transformOTLP(mergedTraceObj)
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/readJsonFile.tsx:34]: Instead you can reassign the result to trace obj and end the if, so that it's further processed by the next if.

**Review by yurishkuro — COMMENTED** [review]
Please add a unit test

File: packages/jaeger-ui/src/utils/readJsonFile.tsx (hunk 1 above)
Comment (Adarsh-jaiss) [reply packages/jaeger-ui/src/utils/readJsonFile.tsx:34]: Sure, I'll do it Asap!

**Comment by Adarsh-jaiss** [comment]
Hey @yurishkuro , I made the signigicant changes : 
```
try {
        let traceObj = JSON.parse(reader.result);
        if (Array.isArray(traceObj) && traceObj.every(obj => 'resourceSpans' in obj)) {
          const mergedResourceSpans = traceObj.reduce((acc, obj) => {
            acc.push(...obj.resourceSpans);
            return acc;
          }, []);

          traceObj = { resourceSpans: mergedResourceSpans };
        }
        if ('resourceSpans' in traceObj) {
          JaegerAPI.transformOTLP(traceObj)
            .then((result: string) => {
              resolve(result);
            })
            .catch(() => {
              reject(new Error(`Error converting traces to OTLP`));
            });
        }
      } catch (error: unknown) {
        reject(new Error(`Error parsing JSON: ${(error as Error).message}`));
      }
    };
```


and added these test cases  : 
```
  it('loads JSON data with single resourceSpans array successfully', async () => {
    const obj = { resourceSpans: [{ id: 1 }, { id: 2 }, { id: 3 }] };
    const file = new File([JSON.stringify(obj)], 'singleResourceSpans.json');
    const result = await readJsonFile({ file });
    expect(result).toEqual(JSON.stringify(obj));
  });

  it('loads JSON data with multiple resourceSpans arrays successfully', async () => {
    const obj1 = { resourceSpans: [{ id: 1 }, { id: 2 }] };
    const obj2 = { resourceSpans: [{ id: 3 }, { id: 4 }] };
    const file = new File([JSON.stringify([obj1, obj2])], 'multipleResourceSpans.json');
    const expectedResult = { resourceSpans: [...obj1.resourceSpans, ...obj2.resourceSpans] };
    const result = await readJsonFile({ file });
    expect(result).toEqual(JSON.stringify(expectedResult));
  });
```

and after running `npm test`, I'm getting this result : 
image

Can you please help me out on this?

**Comment by yurishkuro** [comment]
please push your changes, I cannot troubleshoot via screenshots

**Comment by Adarsh-jaiss** [comment]
- Hey @yurishkuro , I have pushed the code, kindly look into it!

File: packages/jaeger-ui/src/utils/readJsonFile.tsx [hunk 2]
```diff
@@ -32,8 +40,6 @@ export default function readJsonFile(fileList: { file: File }): Promise<string>
             .catch(() => {
               reject(new Error(`Error converting traces to OTLP`));
             });
-        } else {
-          resolve(traceObj);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/readJsonFile.tsx:36]: why is this removed?

Comment (Adarsh-jaiss) [reply packages/jaeger-ui/src/utils/readJsonFile.tsx:36]: I removed it during debugging!!
Adding it now!

**Comment by yurishkuro** [comment]
`yarn lint` and `yarn test` are failing

File: packages/jaeger-ui/src/utils/readJsonFile.tsx [hunk 3]
```diff
@@ -32,7 +40,8 @@ export default function readJsonFile(fileList: { file: File }): Promise<string>
             .catch(() => {
               reject(new Error(`Error converting traces to OTLP`));
             });
-        } else {
+        }
+        else{
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/readJsonFile.tsx:44]:
```suggestion
        } else{
```

File: packages/jaeger-ui/src/utils/readJsonFile.test.js [hunk 4]
```diff
@@ -72,4 +72,20 @@ describe('fileReader.readJsonFile', () => {
     const p = readJsonFile({ file });
     return expect(p).rejects.toMatchObject(expect.any(Error));
   });
+
+  it('loads JSON data with single resourceSpans array successfully', async () => {
+    const obj = { resourceSpans: [{ id: 1 }, { id: 2 }, { id: 3 }] };
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/readJsonFile.test.js:77]: this is not a valid OTEL structure, it should be something like `resourceSpans > scopeSpans > spans`

File: packages/jaeger-ui/src/utils/readJsonFile.test.js [hunk 5]
```diff
@@ -72,4 +72,20 @@ describe('fileReader.readJsonFile', () => {
     const p = readJsonFile({ file });
     return expect(p).rejects.toMatchObject(expect.any(Error));
   });
+
+  it('loads JSON data with single resourceSpans array successfully', async () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/readJsonFile.test.js:76]: I don't think this test is needed, if you can process array of size>1 you can certainly process size=1

File: packages/jaeger-ui/src/utils/readJsonFile.test.js [hunk 6]
```diff
@@ -72,4 +72,20 @@ describe('fileReader.readJsonFile', () => {
...
+    const obj = { resourceSpans: [{ id: 1 }, { id: 2 }, { id: 3 }] };
+    const file = new File([JSON.stringify(obj)], 'singleResourceSpans.json');
+    const result = await readJsonFile({ file });
+    expect(result).toEqual(JSON.stringify(obj));
+  });
+
+  it('loads JSON data with multiple resourceSpans arrays successfully', async () => {
+    const obj1 = { resourceSpans: [{ id: 1 }, { id: 2 }] };
+    const obj2 = { resourceSpans: [{ id: 3 }, { id: 4 }] };
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/readJsonFile.test.js:85]: this is not a valid OTEL structure, it should be something like `resourceSpans > scopeSpans > spans`

**Comment by yurishkuro** [comment]
addressed in #2380
